### PR TITLE
#1415 Evaluation if `dependencies` directory exists

### PR DESCRIPTION
**Pull request by denis256**
Updated terragrunt validation to check if `dependencies` directories exist

Example:
```
$ tree
.
├── module-a
│   ├── main.tf
│   └── terragrunt.hcl
└── terragrunt.hcl

$ cat terragrunt.hcl 
dependencies {
    paths = [
        "../hl3-release",
        "../module-a",
    ]
}

$ terragrunt init
ERRO[0000] Not found dependencies: [../hl3-release (/project/test/hl3-release)] 
ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1

``` 
Included changes:
 * verification of dependencies in `config/config.go`
 * integration test for verification of errors throwing in case of missing module
 * removed from existing tests reference to not existing dependency (test/fixture-read-config/full/source.hcl)
 * updated existing tests to reference existing dependencies;

Fix for issue: https://github.com/gruntwork-io/terragrunt/issues/1415

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -825,6 +829,26 @@ func convertToTerragruntConfig(
...
 }
 
+// Iterate over dependencies paths and check if directories exists, return error with all missing dependencies
+func validateDependencies(terragruntOptions *options.TerragruntOptions, dependencies *ModuleDependencies) error {
+	var missingDependencies []string
+	if dependencies == nil {
+		return nil
+	}
+	for _, dependencyPath := range dependencies.Paths {
+		fullPath := path.Join(terragruntOptions.WorkingDir, dependencyPath)
```
Comment (brikis98) [inline config/config.go:839]: Do you need a check here that `dependencyPath` isn't already an absolute path?

File: config/config.go [hunk 2]
```diff
@@ -825,6 +829,26 @@ func convertToTerragruntConfig(
...
+		fileInfo, err := os.Stat(fullPath)
```
Comment (brikis98) [inline config/config.go:840]: NIT: you can use [`FileExists`](https://github.com/gruntwork-io/terragrunt/blob/master/util/file.go#L44).

File: config/config.go [hunk 3]
```diff
@@ -825,6 +829,26 @@ func convertToTerragruntConfig(
...
+		if err != nil || !fileInfo.IsDir() {
+			missingDependencies = append(missingDependencies, fmt.Sprintf("%s (%s)", dependencyPath, fullPath))
+		}
+	}
+	if len(missingDependencies) != 0 {
```
Comment (brikis98) [inline config/config.go:845]:
```suggestion
	if len(missingDependencies) > 0 {
```

File: config/config.go [hunk 4]
```diff
@@ -893,3 +917,13 @@ func (err InvalidMergeStrategyType) Error() string {
...
 	)
 }
+
+type DependencyDirNotFound struct {
+	Dir []string
+}
+
+func (err DependencyDirNotFound) Error() string {
+	return fmt.Sprintf(
+		"Not found dependencies: %v", err.Dir,
```
Comment (brikis98) [inline config/config.go:927]:
```suggestion
		"Found paths in the 'dependencies' block that do not exist:: %v", err.Dir,
```

**Review by brikis98 — APPROVED** [review]
LGTM! I'll kick off tests now.

**Comment by brikis98** [comment]
Same tests failures here. See https://github.com/gruntwork-io/terragrunt/pull/1795#issuecomment-913626739 for next steps.

**Comment by brikis98** [comment]
OK, https://github.com/gruntwork-io/terragrunt/pull/1799 is now merged. Could you pull in the latest from `master`?

**Comment by denis256** [comment]
Updated with changes from  #1799

**Comment by brikis98** [comment]
Great, thanks! Re-running tests now.

**Comment by brikis98** [comment]
Is this a legit test failure?

```
=== CONT  TestReadTerragruntConfigFull
    integration_test.go:3183: 
        	Error Trace:	integration_test.go:3183
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}(nil)
        	            	actual  : map[string]interface {}{"paths":[]interface {}{"../module-a"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,6 @@
        	            	-(map[string]interface {}) <nil>
        	            	+(map[string]interface {}) (len=1) {
        	            	+ (string) (len=5) "paths": ([]interface {}) (len=1) {
        	            	+  (string) (len=11) "../module-a"
        	            	+ }
        	            	+}
        	            	 
        	Test:       	TestReadTerragruntConfigFull
--- FAIL: TestReadTerragruntConfigFull (19.23s)
```

**Comment by denis256** [comment]
Fixed, now in tests should be referenced existing dependencies

**Review by brikis98 — APPROVED** [review]
Heh, nice, looks like your code is working 😁 

I'll re-run the tests.

**Comment by denis256** [comment]
Fixed tests after merge conflict

**Review by brikis98 — APPROVED** [review]
Re-running!

**Comment by brikis98** [comment]
Alright, tests passed! Merging now.

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.31.10

**Pull request merged**